src/Symfony/Component/DependencyInjection/Compiler/AutowireAnnotatedArgumentsPass.php

```diff
@@ -0,0 +1,141 @@
+
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Looks for definitions with autowiring enabled and parses their "@param" annotations for service ids and parameters.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class AutowireAnnotatedArgumentsPass extends AbstractRecursivePass
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, $isRoot = false)
+    {
+        $value = parent::processValue($value, $isRoot);
+
+        if (!$value instanceof Definition || !$value->isAutowired() || $value->isAbstract() || !$value->getClass()) {
+            return $value;
+        }
+        if (!$reflectionClass = $this->container->getReflectionClass($value->getClass(), false)) {
+            return $value;
+        }
+
+        try {
+            if ($constructor = $this->getConstructor($value, false)) {
+                $constructor = strtolower($constructor->name);
+            }
+        } catch (RuntimeException $e) {
+            return $value;
+        }
+        if (!$annotatedParams = $this->getAnnotatedParams($reflectionClass, $constructor, $value)) {
+            return $value;
+        }
+        $methodCalls = $value->getMethodCalls();
+
+        if ($constructor) {
+            array_unshift($methodCalls, array($constructor, $value->getArguments()));
+        }
+
+        foreach ($methodCalls as $i => $call) {
+            list($method, $arguments) = $call;
+            if (!isset($annotatedParams[$m = strtolower($method)])) {
+                continue;
+            }
+
+            foreach ($annotatedParams[$m] as $index => $param) {
+                if (!array_key_exists($index, $arguments) || ('' === $arguments[$index] && $param instanceof Reference)) {
+                    $arguments[$index] = $param;
+                }
+            }
+            ksort($arguments);
+
+            if ($arguments !== $call[1]) {
+                $methodCalls[$i][1] = $arguments;
+            }
+        }
+
+        if ($constructor) {
+            list(, $arguments) = array_shift($methodCalls);
+
+            if ($arguments !== $value->getArguments()) {
+                $value->setArguments($arguments);
+            }
+        }
+
+        if ($methodCalls !== $value->getMethodCalls()) {
+            $value->setMethodCalls($methodCalls);
+        }
+
+        return $value;
+    }
+
+    private function getAnnotatedParams(\ReflectionClass $reflectionClass, ?string $constructor, Definition $definition): array
+    {
+        $annotatedParams = array();
+
+        if (null !== $constructor) {
+            $annotatedParams[$constructor] = array();
+        }
+        foreach ($definition->getMethodCalls() as list($method)) {
+            $annotatedParams[strtolower($method)] = array();
+        }
+
+        foreach ($reflectionClass->getMethods() as $reflectionMethod) {
+            $r = $reflectionMethod;
```

Comment:
Please, let's not use 1-letter var names 🙏  Thanks!

`$r` -> `$method` or `$reflectedMethod`
`$m` -> `$methodName`
`$p` -> `$parameter` or `$argument` or `$methodParam`